server/src/main/java/org/apache/iotdb/db/query/reader/resourceRelated/UnseqResourceMergeReader.java

```diff
@@ -42,136 +42,133 @@
             .get(tsFileResource, tsFileResource.isClosed());
-        chunkLoader = new ChunkLoaderImpl(tsFileReader);
-      }
+        ChunkLoaderImpl chunkLoader = new ChunkLoaderImpl(tsFileReader);
 
-      for (ChunkMetaData chunkMetaData : metaDataList) {
-
-        if (timeFilter != null && !timeFilter.satisfy(chunkMetaData.getStatistics())) {
-          continue;
+        for (ChunkMetaData chunkMetaData : tsFileMetaDataList) {
+          if (timeFilter != null && !timeFilter.satisfy(chunkMetaData.getStatistics())) {
+            tsFileMetaDataList.remove(chunkMetaData);
+            continue;
+          }
+          chunkMetaData.setPriority(priority++);
+          chunkMetaData.setChunkLoader(chunkLoader);
         }
-        // create and add DiskChunkReader
-        readerWrapList.add(new ChunkReaderWrap(chunkMetaData, chunkLoader, timeFilter));
+        metaDataList.addAll(tsFileMetaDataList);
       }
 
+      // create and add MemChunkReader
       if (!tsFileResource.isClosed()) {
-        // create and add MemChunkReader
-        readerWrapList.add(new ChunkReaderWrap(tsFileResource.getReadOnlyMemChunk(), timeFilter));
+        ChunkReaderWrap memChunkReaderWrap = new ChunkReaderWrap(
+            tsFileResource.getReadOnlyMemChunk(), timeFilter);
+        priorityMergeReader.addReaderWithPriority(memChunkReaderWrap.getIPointReader(), priority++);
       }
     }
 
-    ExternalSortJobEngine externalSortJobEngine = SimpleExternalSortEngine.getInstance();
-    List<IPointReader> readerList = externalSortJobEngine
-        .executeForIPointReader(queryId, readerWrapList);
-    int priorityValue = 1;
-    for (IPointReader chunkReader : readerList) {
-      addReaderWithPriority(chunkReader, priorityValue++);
-    }
+    // sort All ChunkMetadata by start time
+    metaDataList = metaDataList.stream().sorted(Comparator.comparing(ChunkMetaData::getStartTime))
+        .collect(Collectors.toList());
   }
 
-
   /**
-   * Zesong Sun !!!
+   * Create a ChunkReader with priority for each ChunkMetadata and put the ChunkReader to
+   * mergeReader one by one
    */
   @Override
   public boolean hasNextBatch() throws IOException {
-    return false;
-  }
+    ChunkMetaData metaData;
+    ChunkReaderWrap diskChunkReaderWrap;
+    if (priorityMergeReader.hasNext()) {
+      metaData = metaDataList.get(index);
+      long nextMetaDataStartTime = metaData.getStartTime();
+
+      // create and add DiskChunkReader
+      while (priorityMergeReader.current().getTime() >= nextMetaDataStartTime) {
+        diskChunkReaderWrap = new ChunkReaderWrap(metaData, metaData.getChunkLoader(), timeFilter);
+        priorityMergeReader
+            .addReaderWithPriority(diskChunkReaderWrap.getIPointReader(), metaData.getPriority());
+        index++;
+        if (index < metaDataList.size()) {
+          metaData = metaDataList.get(index);
+          nextMetaDataStartTime = metaData.getStartTime();
```

Comment:
What if here `index = metaDataList.size()` and the last updated `nextMetaDataStartTime` <= `priorityMergeReader.current()`, will the `while (priorityMergeReader.current().getTime() >= nextMetaDataStartTime)` loop forever?

Reply:
Hi, thanks for your reminding! I have modified the code logic in a better way.